apps/openmw/mwmechanics/aicombat.cpp

```diff
@@ -326,95 +284,66 @@ namespace MWMechanics
         if (currentAction.get())
             currentAction->getCombatRange(rangeAttack, rangeFollow);
 
-        // FIXME: consider moving this stuff to ActionWeapon::getCombatRange
         const ESM::Weapon *weapon = NULL;
-        MWMechanics::WeaponType weaptype = WeapType_None;
-        float weapRange = 1.0f;
+        MWMechanics::WeaponType weapType = MWMechanics::WeaponType::WeapType_None;
+        float spellSpeed = 1.f;
 
-        // Get weapon characteristics
-        if (actorClass.hasInventoryStore(actor))
+        if (dynamic_cast<ActionWeapon*>(currentAction.get()))
         {
-            // TODO: Check equipped weapon and equip a different one if we can't attack with it
-            // (e.g. no ammunition, or wrong type of ammunition equipped, etc. autoEquip is not very smart in this regard))
+            weapon = dynamic_cast<ActionWeapon*>(currentAction.get())->getWeapon();
 
-            //Get weapon speed and range
-            MWWorld::ContainerStoreIterator weaponSlot =
-                MWMechanics::getActiveWeapon(actorClass.getCreatureStats(actor), actorClass.getInventoryStore(actor), &weaptype);
+            if (weapon != NULL)
+                weapType = getWeaponType(static_cast<ESM::Weapon::Type>(weapon->mData.mType));
+            else
+                weapType = WeapType_HandToHand;
 
-            if (weaptype == WeapType_HandToHand)
-            {
-                static float fHandToHandReach =
-                    world->getStore().get<ESM::GameSetting>().find("fHandToHandReach")->getFloat();
-                weapRange = fHandToHandReach;
-            }
-            else if (weaptype != WeapType_PickProbe && weaptype != WeapType_Spell && weaptype != WeapType_None)
+            if ((weapon == NULL && storage.mLastWeapon != "handToHand") 
+                || (weapon != NULL && weapon->mModel != storage.mLastWeapon))
             {
-                // All other WeapTypes are actually weapons, so get<ESM::Weapon> is safe.
-                weapon = weaponSlot->get<ESM::Weapon>()->mBase;
-                weapRange = weapon->mData.mReach;
+                storage.updateWeaponAttackDurations(actor, weapType, weapon == NULL ? 1.f : weapon->mData.mSpeed, ESM::RT_Self);
```

Comment:
> Anyway there will be rather unpleasant extension of AiSequence::execute, where you'll need to check if package is AiCombat and perform search for that CharacterController on every execution, so specifically for this maybe better would be to queue world state to find CharacterController in Aicombat constructor.

You can't assign a CharacterController in the AiCombat constructor. AiCombat just holds the package information, but can't store any information about the active cells. The AiCombat package persist even when the actor leaves the active cells (at least for the moment, which I plan on reworking for #2172).

As for passing the characterController to execute, that's simple:

<pre>
void AiSequence::execute (const MWWorld::Ptr& actor, AiState& state, CharacterController* character, float duration)
...
            if (package->execute (actor,state,character,duration))
</pre>

Actors::update:
<pre>
                            if (isConscious(iter->first))
                                stats.getAiSequence().execute(iter->first,iter->second->getAiState(), iter->second->getCharacterController(), duration);
</pre>

From what I see the entire getMinMaxAttackDurations function is then not necessary. That's a whole 70 lines of duplicated logic removed. Sizable improvement, IMO.